src/sorting/countingsort.lua

```diff
@@ -0,0 +1,58 @@
+function counting_sort(arr, key_function)
```

Comment:
1. This function should be `return`ed directly for `require` to work with this as expected rather than polluting globals;
2. I would rename `arr` to `list` for consistency with the rest of the code.
3. `key_function` should have a comment explaining what it does and what it defaults to. Especially since this sorting algorithm has special requirements (keys must be integers).